clientv3/client.go

```diff
@@ -280,9 +291,60 @@ func newClient(cfg *Config) (*Client, error) {
 		logger.Set(log.New(ioutil.Discard, "", 0))
 	}
 
+	if cfg.AutoSync > time.Duration(0) {
+		go client.autoSync()
+	}
 	return client, nil
 }
 
+func (c *Client) autoSync() {
+	if c.cfg.AutoSync == time.Duration(0) {
+		return
+	}
+	for {
+		select {
+		case <-time.After(c.cfg.AutoSync):
+			c.sync()
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Client) sync() {
+	mresp, err := c.MemberList(c.ctx)
+	if err != nil {
+		logger.Println(err)
+		return
+	}
+	epm := make(map[string]struct{})
+	for _, m := range mresp.Members {
+		for _, ep := range m.ClientURLs {
+			epm[ep] = struct{}{}
+		}
+	}
+	var eps []string
+	for ep := range epm {
+		eps = append(eps, ep)
+	}
+
+	c.balancer.mu.Lock()
```

Comment:
only balancer methods should lock the balancer so there's a clear division of responsibility